<a href="https://colab.research.google.com/github/jswap21/AG/blob/main/Copy_of_AT_Lesson_40_Reference_v0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 40: Air Quality Analysis - Data Cleaning

**WARNING:** The reference notebook is meant **ONLY** for a teacher. Please **DO NOT** share it with any student. The contents of the reference notebook are meant only to prepare a teacher for a class. To conduct the class, use the class copy of the reference notebook. 

|Particulars|Description|
|-|-|
|**Topic**|Air Quality Analysis - The `datetime` Module|
|||
|**Class Description**|In this class, a student will learn to clean a DataFrame by treating garbage value|
|||
|**Class**|C40|
|||
|**Class Time**|45 minutes|
|||
|**Goals**|Replace the values containing comma with periods|
||Replace the `-200` value with the median values for each column|
||Slice a DataFrame into two DataFrames|
|||
|**Teacher Resources**|Google Account|
||Link to Lesson 40 Colab reference notebook|
||Laptop with internet connectivity|
||Earphones with mic|
|||
|**Student Resources**|Google Account|
||Laptop with internet connectivity|
||Earphones with mic|

---

### Warm-up
```
TEACHER
I have an exciting quiz question for you! Are you ready to answer this question?

EXPECTED STUDENT RESPONSE
Yes.
```

  **Instructions for the Teacher:**
  -  Please click on the "Quiz Time" button on the bottom right corner of your screen to start the In-Class Quiz.<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/quiz-time.png' width=100>   
  - A quiz will be visible to both you and the student. Encourage the student to answer the quiz question. 
  - The student may choose the wrong option, help the student to think correctly about the question and then answer again. 
  - After the student selects the correct option, the "End Quiz" button will start appearing on your screen.  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/end-quiz.png' width=100>  
  - Click the "End quiz" to close the quiz pop-up and continue the class. 
  - Do not spend more than 2 minutes on this quiz.




---

### Teacher-Student Activities

In this class, we will clean the DataFrame by replacing commas with periods (or dots) and by replacing the `-200` garbage value with the median value of the respective columns.

The data cleaning exercises depend on the kind of dataset you get. However, the functions applied or the operations followed to clean a DataFrame more or less remain the same. 

Let's begin the class by running the code which we have already covered in the previous classes and continue the class from the **Activity 1: Replacing Commas** section.

---

### Recap

Run the code cell below to load the dataset and apply the operations on the DataFrame that were covered in the previous classes.

In [ ]:
# Run the code cell.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Loading the dataset.
csv_file = 'https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/air-quality/AirQualityUCI.csv'
df = pd.read_csv(csv_file, sep=';')

# Dropping the 'Unnamed: 15' & 'Unnamed: 16' columns.
df = df.drop(columns=['Unnamed: 15', 'Unnamed: 16'], axis=1) 

# Dropping the null values.
df = df.dropna()

# Creating a Pandas series containing 'datetime' objects.
dt_series = pd.Series(data = [item.split("/")[2] + "-" + item.split("/")[1] + "-" + item.split("/")[0] for item in df['Date']], index=df.index) + ' ' + pd.Series(data=[str(item).replace(".", ":") for item in df['Time']], index=df.index)
dt_series = pd.to_datetime(dt_series)

# Remove the Date & Time columns from the DataFrame and insert the 'dt_series' in it.
df = df.drop(columns=['Date', 'Time'], axis=1)
df.insert(loc=0, column='DateTime', value=dt_series)

# Get the Pandas series containing the year values as integers.
year_series = dt_series.dt.year

# Get the Pandas series containing the month values as integers.
month_series = dt_series.dt.month

# Get the Pandas series containing the day values as integers.
day_series = dt_series.dt.day

# Get the Pandas series containing the days of a week, i.e., Monday, Tuesday, Wednesday etc.
day_name_series = dt_series.dt.day_name()

# Add the 'Year', 'Month', 'Day' and 'Day Name' columns to the DataFrame.
df['Year'] = year_series
df['Month'] = month_series
df['Day'] = day_series
df['Day Name'] = day_name_series

# Sort the DataFrame by the 'DateTime' values in the ascending order. Also, display the first 10 rows of the DataFrame.
df = df.sort_values(by='DateTime')

---

#### Activity 1: Replacing Commas^

The  values in the `CO(GT), C6H6(GT), T,	RH` and	`AH` columns contain the commas in them. Let's replace the commas with periods (or dots). 

In [ ]:
# S1.1: Display the first five rows of the DataFrame.
df.head()

DateTime CO(GT)  PT08.S1(CO)  ...  Month Day   Day Name
0 2004-03-10 18:00:00    2,6       1360.0  ...      3  10  Wednesday
1 2004-03-10 19:00:00      2       1292.0  ...      3  10  Wednesday
2 2004-03-10 20:00:00    2,2       1402.0  ...      3  10  Wednesday
3 2004-03-10 21:00:00    2,2       1376.0  ...      3  10  Wednesday
4 2004-03-10 22:00:00    1,6       1272.0  ...      3  10  Wednesday

[5 rows x 18 columns]

Create a function which takes a Pandas series containing comma separated decimal values as an input and returns a new Pandas series containing period separated decimal values as an output.

In [ ]:
# S1.2: Create a function to replace the commas with periods in a Pandas series.
def comma_to_period(series):
    new_series = pd.Series(data=[float(str(item).replace(',', '.')) for item in series], index=df.index)
    return new_series

In the above code,

- we iterated through each item of a `series`, 

- converted each item to a string value using the `str()` function, 

- replaced the comma with the period using the `replace()` function 

- converted the replaced value to a floating-point number using the `float()` function,

- added each item to a Python list, and

- converted the list to a Pandas series using the `pd.Series()` function. 

We have ensured that the indices of the new Pandas series are the same as the indices of the DataFrame by passing `index = df.index` parameter to the `pd.Series()` function where `df` is the DataFrame.

Now, let's test the `comma_to_period()` function by correcting the values of the `CO(GT)` column but storing the output in a new variable so that the original values of the column are unaffected before we have verified the correctness of the function.

In [ ]:
# S1.3: Test the 'comma_to_period()' function by correcting the values of 'CO(GT)' column but storing the output in a new variable.
new_series = comma_to_period(df['CO(GT)'])
new_series

0       2.6
1       2.0
2       2.2
3       2.2
4       1.6
       ... 
9352    3.1
9353    2.4
9354    2.4
9355    2.1
9356    2.2
Length: 9357, dtype: float64

As you can see, all the values of the `CO(GT)` column are now proper float-point numbers. Let's apply the `comma_to_period()` function on the `'CO(GT), C6H6(GT), T, RH` and `AH` columns.


In [ ]:
# S1.4: Apply the 'comma_to_period()' function on the ''CO(GT)', 'C6H6(GT)', 'T', 'RH' and 'AH' columns.
cols_to_correct = ['CO(GT)', 'C6H6(GT)', 'T', 'RH', 'AH'] # Create a list of column names.
for col in cols_to_correct: # Iterate through each column
    df[col] = comma_to_period(df[col]) # Replace the original column with the new series.

In the above code:

- We created a list of all the columns whose values we need to correct.

- Iterated through each value of the list.

- Applied the `comma_to_period()` function on each column of the `df` DataFrame.

- Replaced each column with their corresponding new series to correct the required values.

Let's print the first five rows of the DataFrame to see the effect of the `comma_to_period()` function.

In [ ]:
# S1.5: Display the first five rows of the DataFrame to see the effect of the 'comma_to_period()' function.
df.head()

DateTime  CO(GT)  PT08.S1(CO)  ...  Month  Day   Day Name
0 2004-03-10 18:00:00     2.6       1360.0  ...      3   10  Wednesday
1 2004-03-10 19:00:00     2.0       1292.0  ...      3   10  Wednesday
2 2004-03-10 20:00:00     2.2       1402.0  ...      3   10  Wednesday
3 2004-03-10 21:00:00     2.2       1376.0  ...      3   10  Wednesday
4 2004-03-10 22:00:00     1.6       1272.0  ...      3   10  Wednesday

[5 rows x 18 columns]

Let's also print the information on the `df` DataFrame. Except for the first column and the last column, all other columns must have the numeric (`float` or `int`) data-type values.

In [ ]:
# S1.6: Print the first information of the 'df' DataFrame.
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   DateTime       9357 non-null   datetime64[ns]
 1   CO(GT)         9357 non-null   float64       
 2   PT08.S1(CO)    9357 non-null   float64       
 3   NMHC(GT)       9357 non-null   float64       
 4   C6H6(GT)       9357 non-null   float64       
 5   PT08.S2(NMHC)  9357 non-null   float64       
 6   NOx(GT)        9357 non-null   float64       
 7   PT08.S3(NOx)   9357 non-null   float64       
 8   NO2(GT)        9357 non-null   float64       
 9   PT08.S4(NO2)   9357 non-null   float64       
 10  PT08.S5(O3)    9357 non-null   float64       
 11  T              9357 non-null   float64       
 12  RH             9357 non-null   float64       
 13  AH             9357 non-null   float64       
 14  Year           9357 non-null   int64         
 15  Month          9357 n

---


#### Activity 2: Garbage Value Inspection^^

The columns in the `df` DataFrame also contain `-200` value. It is a garbage value or just a random number to represent the further missing (or null) values in the DataFrame. Let's replace it with the most appropriate values for each column.

In [ ]:
# S2.1: Get the descriptive statistics for all the numeric data-type columns.
df.describe()

CO(GT)  PT08.S1(CO)  ...        Month          Day
count  9357.000000  9357.000000  ...  9357.000000  9357.000000
mean    -34.207524  1048.990061  ...     6.310356    15.876884
std      77.657170   329.832710  ...     3.438160     8.808653
min    -200.000000  -200.000000  ...     1.000000     1.000000
25%       0.600000   921.000000  ...     3.000000     8.000000
50%       1.500000  1053.000000  ...     6.000000    16.000000
75%       2.600000  1221.000000  ...     9.000000    23.000000
max      11.900000  2040.000000  ...    12.000000    31.000000

[8 rows x 16 columns]

Except for the `Year, Month` and `Day` columns all the other columns contain the `-200` as its minimum value. Because of this their mean values are also affected. The median value never gets affected due to some garbage value or very high and very low values. So, we can replace `-200` with the median value for each column. 

But first, let's find out how many rows contain `-200` in each column except for the `DateTime, Year, Month` and `Day` columns.

In [ ]:
# T2.1: How many rows contain -200 in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns?
neg_counts = [(col, df[df[col] == -200].shape[0]) for col in df.columns[1:-4]]
neg_counts

[('CO(GT)', 1683),
 ('PT08.S1(CO)', 366),
 ('NMHC(GT)', 8443),
 ('C6H6(GT)', 366),
 ('PT08.S2(NMHC)', 366),
 ('NOx(GT)', 1639),
 ('PT08.S3(NOx)', 366),
 ('NO2(GT)', 1642),
 ('PT08.S4(NO2)', 366),
 ('PT08.S5(O3)', 366),
 ('T', 366),
 ('RH', 366),
 ('AH', 366)]

In the above code:

- we iterated through each column having indices between `1` and `-4` (excluding `-4`) in the `df` DataFrame,

- retrieved the rows containing `-200` value using the `df[df[col] == -200]` operation,

- calculated the number of such rows using the `shape[0]` attribute, and

- added the column name and the corresponding counts of `-200` in each column as a tuple in a list using the list comprehension method.

Now, let's also find out the percentage of rows containing `-200` in each column except for the `DateTime, Year, Month` and `Day` columns.

In [ ]:
# S2.2: Find out the percentage of rows containing -200 in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
neg_percent = [(col, round(df[df[col] == -200].shape[0] * 100 / df.shape[0], 2)) for col in df.columns[1:-4]]
neg_percent

[('CO(GT)', 17.99),
 ('PT08.S1(CO)', 3.91),
 ('NMHC(GT)', 90.23),
 ('C6H6(GT)', 3.91),
 ('PT08.S2(NMHC)', 3.91),
 ('NOx(GT)', 17.52),
 ('PT08.S3(NOx)', 3.91),
 ('NO2(GT)', 17.55),
 ('PT08.S4(NO2)', 3.91),
 ('PT08.S5(O3)', 3.91),
 ('T', 3.91),
 ('RH', 3.91),
 ('AH', 3.91)]

Using the same process as above, we calculate the percentage of rows containing `-200` in each column except for the `DateTime, Year, Month` and `Day` columns.

Now, let's remove all the columns from the `df` DataFrame containing more than 10% garbage value.

In [ ]:
# S2.3: Remove all the columns from the 'df' DataFrame containing more than 10% garbage value.
df = df.drop(columns=['NMHC(GT)', 'CO(GT)', 'NOx(GT)', 'NO2(GT)'], axis=1)
df.columns

Index(['DateTime', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)', 'PT08.S3(NOx)',
       'PT08.S4(NO2)', 'PT08.S5(O3)', 'T', 'RH', 'AH', 'Year', 'Month', 'Day',
       'Day Name'],
      dtype='object')

Again, calculate the percentage of rows containing `-200` in each column except for the `DateTime, Year, Month` and `Day` columns.

In [ ]:
# S2.4: Again, calculate the percentage of rows containing '-200' in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
neg_percent = [(col, round(df[df[col] == -200].shape[0] * 100 / df.shape[0], 2)) for col in df.columns[1:-4]]
neg_percent

[('PT08.S1(CO)', 3.91),
 ('C6H6(GT)', 3.91),
 ('PT08.S2(NMHC)', 3.91),
 ('PT08.S3(NOx)', 3.91),
 ('PT08.S4(NO2)', 3.91),
 ('PT08.S5(O3)', 3.91),
 ('T', 3.91),
 ('RH', 3.91),
 ('AH', 3.91)]

Let's replace the `-200` value with the median values in all the columns except for the `DateTime, Year, Month` and `Day` columns. 

Before that we should split the entire DataFrame in two different DataFrames because it contains data for two different years, i.e., 2004 and 2005. Then we should calculate the median values for each column for 2004 and 2005 separately. However, we can also first find out whether the median values are actually different for the two years. If they are not, then we don't need to split the DataFrame into two DataFrames for 2004 and 2005 data points.

In [ ]:
# T2.2: Calculate the median values for the columns having indices between 1 and -4 (excluding -4) for the year 2004.
df.loc[df['Year'] == 2004, df.columns[1:-4]].describe()

PT08.S1(CO)     C6H6(GT)  ...           RH           AH
count  7110.000000  7110.000000  ...  7110.000000  7110.000000
mean   1055.135443     3.876779  ...    40.292053    -5.310343
std     314.719517    37.859514  ...    46.983126    35.441047
min    -200.000000  -200.000000  ...  -200.000000  -200.000000
25%     919.000000     4.600000  ...    33.000000     0.855925
50%    1052.000000     8.600000  ...    47.700000     1.083550
75%    1221.000000    14.200000  ...    61.000000     1.404175
max    2040.000000    63.700000  ...    88.700000     2.231000

[8 rows x 9 columns]

In [ ]:
# S2.5: Calculate the median values for the columns having indices between 1 and -4 (excluding -4) for the year 2005.
df.loc[df['Year'] == 2005, df.columns[1:-4]].describe()

PT08.S1(CO)     C6H6(GT)  ...           RH           AH
count  2247.000000  2247.000000  ...  2247.000000  2247.000000
mean   1029.544726    -4.497864  ...    36.932888   -11.670190
std     373.059005    50.426047  ...    62.700754    48.186199
min    -200.000000  -200.000000  ...  -200.000000  -200.000000
25%     925.500000     2.900000  ...    37.200000     0.413350
50%    1055.000000     5.800000  ...    50.900000     0.596400
75%    1222.000000    11.200000  ...    64.750000     0.804950
max    1846.000000    43.000000  ...    86.600000     1.393000

[8 rows x 9 columns]

As we can see, the median values for 2004 and 2005 are different. Hence, we should split the `df` DataFrame into two different DataFrames. One for 2004 data points and another for 2005 data points.

---

#### Activity 3: Garbage Value Replacement^^^

Before we separate the DataFrame, let's calculate the count of 2004 and 2005 data points.

In [ ]:
# S3.1: Count the number of rows containing '2004' and '2005' year values.
df['Year'].value_counts()

2004    7110
2005    2247
Name: Year, dtype: int64

The DataFrame for 2004 year should have 7110 rows which means the DataFrame for 2005 year should have 2247 rows.

In [ ]:
# S3.2: Create a new DataFrame containing records for the year 2004. Also, display the first five rows.
aq_2004_df = df[df['Year'] == 2004]
aq_2004_df.head()

DateTime  PT08.S1(CO)  C6H6(GT)  ...  Month  Day   Day Name
0 2004-03-10 18:00:00       1360.0      11.9  ...      3   10  Wednesday
1 2004-03-10 19:00:00       1292.0       9.4  ...      3   10  Wednesday
2 2004-03-10 20:00:00       1402.0       9.0  ...      3   10  Wednesday
3 2004-03-10 21:00:00       1376.0       9.2  ...      3   10  Wednesday
4 2004-03-10 22:00:00       1272.0       6.5  ...      3   10  Wednesday

[5 rows x 14 columns]

In [ ]:
# S3.3 Calculate the number of rows and columns in the DataFrame for the 2004 year records.
aq_2004_df.shape

(7110, 14)

In [ ]:
# S3.4: Create a new DataFrame containing records for the year 2005. Also, display the first five rows.
aq_2005_df = df[df['Year'] == 2005]
aq_2005_df.head()

DateTime  PT08.S1(CO)  C6H6(GT)  ...  Month  Day  Day Name
7110 2005-01-01 00:00:00       1046.0       4.2  ...      1    1  Saturday
7111 2005-01-01 01:00:00       1275.0       8.8  ...      1    1  Saturday
7112 2005-01-01 02:00:00       1173.0       7.5  ...      1    1  Saturday
7113 2005-01-01 03:00:00       1163.0       7.6  ...      1    1  Saturday
7114 2005-01-01 04:00:00       1054.0       5.6  ...      1    1  Saturday

[5 rows x 14 columns]

In [ ]:
# S3.5 Calculate the number of rows and columns in the DataFrame for the 2004 year records.
aq_2005_df.shape

(2247, 14)

Now, let's replace the `-200` value with the median values for each column having indices between 1 and -4 (excluding -4) for both the 2004 and 2005 year DataFrames.

In [ ]:
# T3.1: Replace the -200 value with the median values for each column having indices between 1 and -4 (excluding -4) for the 2004 year DataFrame.
import warnings
warnings.filterwarnings('ignore')

for col in aq_2004_df.columns[1:-4]:
  median = aq_2004_df.loc[aq_2004_df[col] != -200, col].median() # Get the median value for each column after excluding -200.
  aq_2004_df[col] = aq_2004_df[col].replace(to_replace=-200, value=median)

```
import warnings
warnings.filterwarnings('ignore')
```

The above code will remove all the warnings that we may get.

In [ ]:
# S3.6: Repeat the same exercise for the 2005 year DataFrame.
for col in aq_2005_df.columns[1:-4]:
  median = aq_2005_df.loc[aq_2005_df[col] != -200, col].median()
  aq_2005_df[col] = aq_2005_df[col].replace(to_replace=-200, value=median)

Now, let's again, calculate the number of rows containing '-200' in each column except for the `DateTime, Year, Month` and `Day` columns in both the DataFrames.

In [ ]:
# S3.7: Compute the number of rows containing '-200' in each column having indices between 1 and -4 (excluding -4) in the 2004 year DataFrame.
neg_val_2004 = [(col, aq_2004_df[aq_2004_df[col] == -200].shape[0]) for col in df.columns[1:-4]]
neg_val_2004

[('PT08.S1(CO)', 0),
 ('C6H6(GT)', 0),
 ('PT08.S2(NMHC)', 0),
 ('PT08.S3(NOx)', 0),
 ('PT08.S4(NO2)', 0),
 ('PT08.S5(O3)', 0),
 ('T', 0),
 ('RH', 0),
 ('AH', 0)]

In [ ]:
# S3.8: Again, calculate the percentage of rows containing '-200' in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
neg_val_2005 = [(col, aq_2005_df[aq_2005_df[col] == -200].shape[0]) for col in df.columns[1:-4]]
neg_val_2005

[('PT08.S1(CO)', 0),
 ('C6H6(GT)', 0),
 ('PT08.S2(NMHC)', 0),
 ('PT08.S3(NOx)', 0),
 ('PT08.S4(NO2)', 0),
 ('PT08.S5(O3)', 0),
 ('T', 0),
 ('RH', 0),
 ('AH', 0)]

Let's pause here. In the next class, we will learn how to group a DataFrame about a particular column to perform aggregation operations such as count, mean, median and sum. 

---

#### Activities

**Teacher Activities**

1. Air Quality Analysis - Data Cleaning (Class Copy)

   https://colab.research.google.com/drive/1oY8-DPgjTYTALDMVRjrdNO7R7YNxw3in
   
2. Air Quality Analysis - Data Cleaning (Reference)

   https://colab.research.google.com/drive/1MMW7IovWOQ61cWTgCqp8O7gMiMI_s5Hm

---